include/envoy/stream_info/stream_info.h

```diff
@@ -520,6 +525,17 @@ class StreamInfo {
    * @return request headers.
    */
   virtual const Http::RequestHeaderMap* getRequestHeaders() const PURE;
+
+  /**
+   * @param Upstream Connection's ClusterInfo.
+   */
+  virtual void
+  setUpstreamClusterInfo(const Upstream::ClusterInfoConstSharedPtr& upstream_cluster_info) PURE;
+
+  /**
+   * @return Upstream Connection's ClusterInfo.
+   */
+  virtual absl::optional<Upstream::ClusterInfoConstSharedPtr> upstreamClusterInfo() const PURE;
```

Comment:
I don't think we need an optional here. Just return the pointer which can be null

Reply:
+ @kyessenov 
@mattklein123 : we felt it would be good to have 3 states here : unset, set to empty, set to clusterinfo, wherein set to empty can tell about an unhealthy host and thus is empty and unset for things like misconfiguration.. wdyt?

Reply:
Sure that's fine but please clarify the API docs.

Reply:
Done